### ISSUE TITLE
[BUG] Doesn't tolerate a restart of Home Assistant when the Sky Q is in Eco Mode

### ISSUE THREAD

**Issue by shuntfield** [RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer#106]
**Describe the bug**  
Doesn't tolerate a restart of Home Assistant when the Sky Q is in Eco Mode, I have any automation that restarts Home Assistant at 4am each day, mainly because I have an issue with a different integration that the dev has ignored, 4am because its dark and very little runs overnight.
I am in the UK using Sky Q - UK.

**To Reproduce**  
1. Allow the boxes to go into Eco Mode overnight
2. Restart Home Assistant when Sky Q is in Eco Mode

**Expected behavior**  
To wait or poll until eco mode has ended and then reattampt to establish communication without throwing an error into the log. 

**Screenshots**  
![image](https://user-images.githubusercontent.com/63152393/169657688-7fb005db-fe6c-4eb0-bd01-f6015368527e.png)
![image](https://user-images.githubusercontent.com/63152393/169657705-f0b8c749-d567-4eec-81a1-0d13b68bd987.png)

**Component versions**  
V2.9.2

**Your configuration**  
How have you setup your devices, by UI or YAML?
UI Setup

**Output of HA logs**  
Paste the relavant output of the HA log here.
```This error originated from a custom integration.

Logger: homeassistant.config_entries
Source: custom_components/skyq/__init__.py:44
Integration: Sky Q (documentation, issues)
First occurred: 04:05:49 (2 occurrences)
Last logged: 04:05:49

Error setting up entry Top's Bedroom Sky Q Mini for skyq
Error setting up entry Top's Gaff Sky Q for skyq
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 335, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/skyq/__init__.py", line 44, in async_setup_entry
    remote = await hass.async_add_executor_job(SkyQRemote, host, epg_cache_len)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/pyskyqremote/skyq_remote.py", line 60, in __init__
    self._setup_remote()
  File "/usr/local/lib/python3.9/site-packages/pyskyqremote/skyq_remote.py", line 332, in _setup_remote
    device_info = self.get_device_information()
  File "/usr/local/lib/python3.9/site-packages/pyskyqremote/skyq_remote.py", line 287, in get_device_information
    self._remote_config.set_device_info(device_info)
  File "/usr/local/lib/python3.9/site-packages/pyskyqremote/skyq_remote.py", line 378, in set_device_info
    self.territory = TERRITORIES[device_info.used_country_code]
AttributeError: 'NoneType' object has no attribute 'used_country_code'
```

**Additional context**  
Great integration by the way, not knocking it at all.
Working around the issue is by way of restart of HA or delete and reinstalling the integration or disabling and then reenabling the entity. It seems the integration doesn't repoll for the Sky Q devices after they come back up.
Reloading the integration from the UI might help, but that option isn't in the list.
The Sky Q comes out of eco around 05:45hrs each day.

**Comment by RogerSelwyn** [comment]
Should be fixed with - https://github.com/RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer/releases/tag/v2.9.3

**Comment by shuntfield** [comment]
Awesome, I have updated and will test overnight, closing if it appears resolved

Great stuff!

**Comment by shuntfield** [comment]
![image](https://user-images.githubusercontent.com/63152393/169677419-cc661f2e-e9d5-49a5-9aa0-71b5636f20ed.png)

OK My HA instance has restarted at 4am, but this error has appeared now, the ECO mode hasn't ended yet, so will wait and see what it does when it comes out of ECO.

**Comment by shuntfield** [comment]
Also in the log:
```
This error originated from a custom integration.

Logger: aiohttp.server
Source: custom_components/skyq/config_flow.py:140
Integration: Sky Q (documentation, issues)
First occurred: 04:36:20 (1 occurrences)
Last logged: 04:36:20

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 94, in forwarded_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 79, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 219, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 137, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 243, in post
    return await super().post(request)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 62, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 70, in post
    result = await self._flow_mgr.async_init(
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 205, in async_init
    flow, result = await task
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 232, in _async_init
    result = await self._async_handle_step(flow, flow.init_step, data, init_done)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 335, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/config/custom_components/skyq/config_flow.py", line 140, in async_step_init
    self._remote = self.hass.data[DOMAIN][self._config_entry.entry_id][SKYQREMOTE]
KeyError: '2542cb303a2495ee69e7dda390e70266'
```

**Comment by RogerSelwyn** [comment]
I'm sure I've seen someone else with this error, which is really strange, because it shouldn't go through the config_flow during a restart. You only exercise that code when you hit configure on the integration panel page.

**Comment by RogerSelwyn** [comment]
Yep found the one that also had this, but I really don't know what is triggering it. Though just wondering if it passes through the flow if it can't setup the integration for some time, can't see why because it won't help. I assume this didn't happen with the version previous to my most recent changes?

**Comment by RogerSelwyn** [comment]
Looking at the code, I can see why it would get that error after a restart during ECO mode, if you went to they and configure it. But it shouldn't pass through that code unless you hit the configure button shown in your screenshot. It isn't possible to re-configure the box while it isn't available since the dialogue is dependant on the box being switched on to retrieve various information (channels/etc).

**Comment by bluenazgul** [comment]
with the latest update i can reproduce this:

> Logger: aiohttp.server
> Source: custom_components/skyq/config_flow.py:140
> Integration: Sky Q ([documentation](https://github.com/RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer), [issues](https://github.com/RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer/issues))
> First occurred: 11:47:40 (1 occurrences)
> Last logged: 11:47:40
> 
> Error handling request
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
>     resp = await request_handler(request)
>   File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
>     resp = await handler(request)
>   File "/usr/local/lib/python3.9/site-packages/aiohttp/web_middlewares.py", line 117, in impl
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 79, in ban_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 219, in auth_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 137, in handle
>     result = await result
>   File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 243, in post
>     return await super().post(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 62, in wrapper
>     result = await method(view, request, *args, **kwargs)
>   File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 70, in post
>     result = await self._flow_mgr.async_init(
>   File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 205, in async_init
>     flow, result = await task
>   File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 232, in _async_init
>     result = await self._async_handle_step(flow, flow.init_step, data, init_done)
>   File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 335, in _async_handle_step
>     result: FlowResult = await getattr(flow, method)(user_input)
>   File "/config/custom_components/skyq/config_flow.py", line 140, in async_step_init
>     self._remote = self.hass.data[DOMAIN][self._config_entry.entry_id][SKYQREMOTE]
> KeyError: '7fcbae4d035e11ebbf31330c9796ec6f'

> Logger: custom_components.skyq.classes.power
> Source: custom_components/skyq/classes/power.py:52
> Integration: Sky Q ([documentation](https://github.com/RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer), [issues](https://github.com/RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer/issues))
> First occurred: 11:45:24 (1 occurrences)
> Last logged: 11:45:24
> 
> W0010 - Device is not available: Sky Q

Strange is, my Sky Q was running at the time i restartet HA - i unable to reconfigure the interation, ill try to full reload it now

**Comment by RogerSelwyn** [comment]
You say you can reproduce - What steps are you following to get that error to occur? What version of SkyQ are you using?

I don't believe re downloading from HACS will make any difference, since nothing changes.

**Comment by bluenazgul** [comment]
> You say you can reproduce - What steps are you following to get that error to occur? What version of SkyQ are you using?
> 
> I don't believe re downloading from HACS will make any difference, since nothing changes. HA restarts will make a difference since it will restart the integration.

Integration 2.9.3
Remote *.12

restart HA didnt fixed it for me, only reinstall

My Sky Q is permant in ECOmode when not in use

**Comment by RogerSelwyn** [comment]
So.. Did this happen when you did the restart, or did do anything else?

**Comment by bluenazgul** [comment]
> So.. Did this happen when you did the restart, or did do anything else?

on restart when SkyQ is in ECO Mode and about 30min after SkyQ goes from StandyBy to ECO Mode

**Comment by RogerSelwyn** [comment]
I simply don't see why anything is initiating the config_flow after 30 minutes of a restart in ECO mode. That is a user initiated action.

**Comment by RogerSelwyn** [comment]
@bluenazgul - How do you get your box to go into ECO during the day. I thought you could only configure that for night time. Just make it easier for me to debug.

**Comment by bluenazgul** [comment]
You can Set eco Mode to permanent - sadly the Box then needs about 3min everytime you Start it

**Comment by RogerSelwyn** [comment]
My box has None, Active, ECO. ECO puts it in low power overnight. The others don’t put it in low pier at all. 

![image](https://user-images.githubusercontent.com/7115783/169692946-41efabd3-a7da-43ba-9e11-7599c86c5d77.jpg)

**Comment by RogerSelwyn** [comment]
I've added some logging into the config_flow code. It would be great if you could copy this file into your SkyQ folder:

https://github.com/RogerSelwyn/Home_Assistant_SkyQ_MediaPlayer/blob/master/custom_components/skyq/config_flow.py

**Comment by bluenazgul** [comment]
> My box has None, Active, ECO. ECO puts it in low power overnight. The others don’t put it in low pier at all.
> 
> ![image](https://user-images.githubusercontent.com/7115783/169692946-41efabd3-a7da-43ba-9e11-7599c86c5d77.jpg)

I also have "ECO Plus" on my Box
![IMG_20220522_133928](https://user-images.githubusercontent.com/171537/169693297-ccf30d3d-fa05-4cca-8aff-5a0a8bf2e95e.jpg)

**Comment by RogerSelwyn** [comment]
Looks like something you can get in Germany but not in the UK. Can you try the config_flow.py update?

**Comment by bluenazgul** [comment]
> Looks like something you can get in Germany but not in the UK. Can you try the config_flow.py update?
just updated the file and restarted HA

**Comment by RogerSelwyn** [comment]
I should add you would need to enable debug logging for SkyQ because I've added one debug statement, but also a couple of error situations as well.

**Comment by RogerSelwyn** [comment]
In theory if it enters the config_flow it should give a D0010 message saying it has done so. Then if it is during ECO mode after a restart a E0010. If HA was started when it went into ECO mode it would give an E0020.

As I said before, this should only ever happen when someone presses 'Configure'. There is nothing (that I know of) in my code that will automatically trigger config.

If you go back to 2.9.0 or earlier, does the problem go away? If it does, I can try and see what has changed, because 2.9.1+ were back end code, which really shouldn't have affected what is happening here.

**Comment by bluenazgul** [comment]
maybe it is coz i try to reload the intergration (using HA Service) every 30seconds once my TV is powered on

**Comment by RogerSelwyn** [comment]
I have no idea what affect that will have. Do you do it all the time? Or just until the box shows as switched on?

**Comment by bluenazgul** [comment]
Only Until the Box return to be on
![Screenshot_20220522_140016_io homeassistant companion android](https://user-images.githubusercontent.com/171537/169694157-d3e4dfd6-be18-4082-a3ae-63d5d0fdb8d1.jpg)
![Screenshot_20220522_135909_io homeassistant companion android](https://user-images.githubusercontent.com/171537/169694158-cd0afea3-cb24-4d4b-860e-5b38928f01c9.jpg)
![Screenshot_20220522_135843_io homeassistant companion android](https://user-images.githubusercontent.com/171537/169694160-4f5bd82a-0e28-4c28-9ad1-9a568d8d5a7f.jpg)

**Comment by bluenazgul** [comment]
```
alias: reload sky q
description: ''
trigger:
  - platform: time_pattern
    seconds: '30'
condition:
  - condition: and
    conditions:
      - condition: device
        type: is_on
        device_id: 2fc641b44f44f3c2087c2ca6314dc80c
        entity_id: switch.medialeiste
        domain: switch
      - condition: state
        entity_id: media_player.sky_q
        state: unavailable
action:
  - service: homeassistant.reload_config_entry
    data: {}
    target:
      entity_id:
        - media_player.sky_q
mode: single

```

**Comment by bluenazgul** [comment]
This is the automation i use to reload the Integration

**Comment by RogerSelwyn** [comment]
Looks like HA retries to load the platform with longer and longer times between retries. Still no idea what it is trying to run the config_flow, I'll take a look at what happens when you do a reload.

**Comment by RogerSelwyn** [comment]
When you do a restart or reload you should get this error in the log, if you are in eco mode during a restart. Nothing else should happen.

`2022-05-22 12:07:38 WARNING (MainThread) [homeassistant.config_entries] Config entry 'Sky Q' for skyq integration not ready yet: None; Retrying in background`

**Comment by RogerSelwyn** [comment]
Just to ask again. Do you get errors if you go back to 2.9.0 (or earlier)? Trying to isolate if this is an HA problem or a SkyQ problem.

**Comment by bluenazgul** [comment]
Even on 2.9.0 an earlier i get this, maybe Sky has something changed in API Code

According to the error log i should get, i get it

`2022-05-22 14:08:13 WARNING (MainThread) [homeassistant.config_entries] Config entry 'Sky Q' for skyq integration not ready yet: None; Retrying in background`

**Comment by RogerSelwyn** [comment]
On any version you should get:

`2022-05-22 14:08:13 WARNING (MainThread) [homeassistant.config_entries] Config entry 'Sky Q' for skyq integration not ready yet: None; Retrying in background`

But you shouldn't get:

```
File "/config/custom_components/skyq/config_flow.py", line 140, in async_step_init
self._remote = self.hass.data[DOMAIN][self._config_entry.entry_id][SKYQREMOTE]
KeyError: '7fcbae4d035e11ebbf31330c9796ec6f'
```

Well you should never get the second one, but I'm trying to understand if you are getting that on 2.9.0

**Comment by RogerSelwyn** [comment]
Looking back, the first time you stated you got this is when you tried 0.3.8 pyskyqremote with SkyQ 2.9.0. That release would have no affect on the SkyQ integration (as far as I believe) since it just changed where it got the epg and images from. It was in 0.3.9 that I stripped out loads of redundant code.

But I did make other changes in 2.9.1, just trying to see if any of that would do something weird.

**Comment by shuntfield** [comment]
Hi, as you might gather I worked nights last night 🤣
I checked it after the boxes exited standby and they error goes away if its own accord.

I've been restarting over night for a few weeks and it's only the last few days it has been faulting so it's possibly a recent change.

**Comment by shuntfield** [comment]
> When you do a restart or reload you should get this error in the log, if you are in eco mode during a restart. Nothing else should happen.
> 
> `2022-05-22 12:07:38 WARNING (MainThread) [homeassistant.config_entries] Config entry 'Sky Q' for skyq integration not ready yet: None; Retrying in background`

Yes I get this on a restart and the red box shown earlier.
After the boxes come out of eco then it must repoll and comes up ok.

**Comment by RogerSelwyn** [comment]
So you don’t get this error? Or at least only when you press config when in eco mode. 

```
File
"/config/custom_components/skyq/config_flow.py", line 140, in async_step_init
self._remote = self.hass.data[DOMAIN][self._config_entry.entry_id][SKYQREMOTE]
KeyError: '7fcbae4d035e11ebbf31330c9796ec6f
```

**Comment by RogerSelwyn** [comment]
> ok.

This is what I would expect to happen. HA repolls every couple of minutes until it can setup the integration, and the Red error will go away.

**Comment by RogerSelwyn** [comment]
> > When you do a restart or reload you should get this error in the log, if you are in eco mode during a restart. Nothing else should happen.
> > `2022-05-22 12:07:38 WARNING (MainThread) [homeassistant.config_entries] Config entry 'Sky Q' for skyq integration not ready yet: None; Retrying in background`
> 
> Yes I get this on a restart and the red box shown earlier. After the boxes come out of eco then it must repoll and comes up ok.

This is what I would expect to happen. HA repolls every couple of minutes until it can setup the integration, and the Red error will go away.

**Comment by RogerSelwyn** [comment]
> Hi, as you might gather I worked nights last night 🤣 I checked it after the boxes exited standby and they error goes away if its own accord.
> 
> I've been restarting over night for a few weeks and it's only the last few days it has been faulting so it's possibly a recent change.

@shuntfield - It should have always had the 'Retrying in background' message. That was something I did in the middle of last year some time. So when you say 'faulting' what do you mean?

**Comment by shuntfield** [comment]
> > Hi, as you might gather I worked nights last night 🤣 I checked it after the boxes exited standby and they error goes away if its own accord.
> > I've been restarting over night for a few weeks and it's only the last few days it has been faulting so it's possibly a recent change.
> 
> @shuntfield - It should have always had the 'Retrying in background' message. That was something I did in the middle of last year some time. So when you say 'faulting' what do you mean?

I don't normally get the Red Box it just sits waiting, and no message in the logs that's perhaps what I mean by faulting.

**Comment by RogerSelwyn** [comment]
Not sure when I implemented that, but should have been like that for quite some time. I just went back and checked against 2.8.9 and it behaves the same. 

The item that concerns me is the config _flow error because it shouldn’t happen unless someone tries to manually (or through some automation I’m not aware of) reconfigure the integration. 

Now I could possibly remove the dependency on access to the box if it is also, but that would remove the ability to configure channels (which are retrieved from the box).  But that makes it more complex and I don’t really think it is too much of a burden to make sure the box is switched on when you try to configure the integration.  Consequently I have added some error having (which will be in the next release) to at least give some feedback that you are doing something that won’t work.

**Comment by RogerSelwyn** [comment]
That said, I think HA core may have added the red highlighting fairly recently. I run dev build in my development environment so it’s been like that for quite some time.

**Comment by shuntfield** [comment]
OK, I wasn't aware that you could use a automation that can reload entities as mentioned earlier in the thread.
But now instead of it just abandoning it and then not restarting of its own accord it actually comes up fine when the box comes out of Eco.
I can live with the red box being there in Eco.

**Comment by RogerSelwyn** [comment]
@bluenazgul - I set my box to reboot at 4am, and it did what I would expect. It gave the error saying that the integration is not ready, and then polled approx every 90 seconds until the box came up, when it connected and ran fine.

I can only think that you reloading the integration every 30 seconds is doing something weird if you are getting an error in the config flow. I'm afraid I've spent quite a few hours trying to replicate it and can't.
